pkgs/development/compilers/flutter/flutter.nix

```diff
@@ -30,47 +30,44 @@
 , nspr
 , nss
 , systemd
+, callPackage
 }:
 let
-  drvName = "flutter-${channel}-${version}";
+  repository = callPackage ./repository.nix {
+    inherit src pname version dart depsSha256;
+  };
+  drvName = "flutter-${version}";
+
   flutter = stdenv.mkDerivation {
     name = "${drvName}-unwrapped";
 
-    src = fetchurl {
-      url =
-        "https://storage.googleapis.com/flutter_infra/releases/${channel}/linux/${filename}";
-      sha256 = sha256Hash;
-    };
-
-    buildInputs = [ git ];
+    buildInputs = [ git repository ];
 
-    inherit patches;
+    inherit src patches;
 
     postPatch = ''
       patchShebangs --build ./bin/
       find ./bin/ -executable -type f -exec patchelf --set-interpreter $(cat $NIX_CC/nix-support/dynamic-linker) {} \;
     '';
 
     buildPhase = ''
-      FLUTTER_ROOT=$(pwd)
-      FLUTTER_TOOLS_DIR="$FLUTTER_ROOT/packages/flutter_tools"
-      SNAPSHOT_PATH="$FLUTTER_ROOT/bin/cache/flutter_tools.snapshot"
-      STAMP_PATH="$FLUTTER_ROOT/bin/cache/flutter_tools.stamp"
-      SCRIPT_PATH="$FLUTTER_TOOLS_DIR/bin/flutter_tools.dart"
-      DART_SDK_PATH="${dart}"
+      export FLUTTER_ROOT="$(pwd)"
+      export FLUTTER_TOOLS_DIR="$FLUTTER_ROOT/packages/flutter_tools"
+      export SCRIPT_PATH="$FLUTTER_TOOLS_DIR/bin/flutter_tools.dart"
 
-      HOME=../.. # required for pub upgrade --offline, ~/.pub-cache
-                 # path is relative otherwise it's replaced by /build/flutter
+      mkdir -p "$out/bin/cache"
+      export SNAPSHOT_PATH="$out/bin/cache/flutter_tools.snapshot"
```

Comment:
Just a note: Quotes are not really required because $out can't contain any character which needs quoting.

Reply:
Yeah, I know, but I did this for consistency.